### PR TITLE
Feature/subscribe

### PR DESCRIPTION
**Pull request by danhanson**
Adam broke my branch #24, so here is the pull request for pushes
When you create a vehicle, you are subscribed to receive route updates for that vehicle

### CODE REVIEW

File: app/io/pathfinder/routing/Router.scala [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+        recalculate()
+    }
+
+    def addCommodity(c: Commodity): Unit = {
+        Logger.info("Adding Commodity To routed collection: "+c)
+        commodities += c
+        recalculate()
+    }
+
+    // Dan wrote an infinite loop here.
```
Comment (aj-michael) [inline app/io/pathfinder/routing/Router.scala:46]: Best part! :+1:

File: conf/evolutions/default/1.sql [hunk 2]
```diff
@@ -15,6 +15,7 @@ create table commodity (
   endLatitude               double not null,
```
Comment (aj-michael) [inline conf/evolutions/default/1.sql]: Please remove this file.

File: app/io/pathfinder/data/EbeanCrudDao.scala [hunk 3]
```diff
@@ -4,7 +4,7 @@ import com.avaje.ebean.Model
 import play.db.ebean.Transactional
 import scala.collection.JavaConversions.asScalaBuffer
 
-abstract class EbeanCrudDao[K,M <: Model](finder: Model.Find[K,M]) extends CrudDao[K,M] {
+class EbeanCrudDao[K,M <: Model](finder: Model.Find[K,M]) extends CrudDao[K,M] {
```
Comment (aj-michael) [inline app/io/pathfinder/data/EbeanCrudDao.scala:7]: Please add a comment explaining the purpose of this class.

File: app/io/pathfinder/data/ObserverDao.scala [hunk 4]
```diff
@@ -0,0 +1,49 @@
+package io.pathfinder.data
+
+import com.avaje.ebean.Model
+import com.avaje.ebean.Model.Find
+
+/**
+ * subclasses can use callbacks to listen to changes to models in the specified dao
+ * @param dao
```
Comment (aj-michael) [inline app/io/pathfinder/data/ObserverDao.scala:8]: For these `@param` tags, please either remove them or add descriptions. Also, please be a little more concrete in your description of the class. I cannot explain what it means from this.

File: app/io/pathfinder/websockets/WebSocketActor.scala [hunk 5]
```diff
@@ -1,9 +1,15 @@
...
+import io.pathfinder.routing.Router
+import io.pathfinder.websockets.ModelTypes.ModelType
+import io.pathfinder.websockets.controllers.VehicleSocketController._
+import play.Logger
 import play.api.libs.concurrent.Execution.Implicits.defaultContext
 import io.pathfinder.websockets.controllers.{CommoditySocketController, ClusterSocketController, WebSocketController, VehicleSocketController}
+import scala.concurrent.ExecutionContext.Implicits.global
+import play.api.libs.iteratee.{Iteratee, Concurrent}
 
 object WebSocketActor {
```
Comment (aj-michael) [inline app/io/pathfinder/websockets/WebSocketActor.scala:14]: Why is this an object and not a class? If all of our web socket requests are routed through a singleton, we are limiting our scaling potential

File: app/io/pathfinder/websockets/WebSocketActor.scala [hunk 6]
```diff
@@ -24,18 +30,48 @@ class WebSocketActor (
     controllers: Map[ModelType, WebSocketController]
 ) extends Actor {
     import WebSocketMessage._
+    Router
```
Comment (aj-michael) [inline app/io/pathfinder/websockets/WebSocketActor.scala:33]: This line should not go here. If we put it here, the Router object is not initialized until the first socket request is made. This causes the first socket connection to be delayed.

File: app/io/pathfinder/websockets/WebSocketActor.scala [hunk 7]
```diff
@@ -1,9 +1,15 @@
 package io.pathfinder.websockets
 
 import akka.actor.{Props, Actor, ActorRef}
-import io.pathfinder.websockets.ModelTypes._
+import io.pathfinder.models.{Vehicle, Commodity}
+import io.pathfinder.routing.Router
+import io.pathfinder.websockets.ModelTypes.ModelType
+import io.pathfinder.websockets.controllers.VehicleSocketController._
```
Comment (aj-michael) [inline app/io/pathfinder/websockets/WebSocketActor.scala:7]: You know my feelings about wildcard imports.

File: app/io/pathfinder/models/Commodity.scala [hunk 8]
```diff
@@ -5,11 +5,12 @@ import javax.persistence.{Column, Entity, GeneratedValue, GenerationType, Id, Ma
 import com.avaje.ebean.Model
 import com.fasterxml.jackson.annotation.JsonIgnore
 import io.pathfinder.data.{Resource, EbeanCrudDao}
+import io.pathfinder.websockets.WebSocketDao
 import play.api.libs.json.{Json, Format}
 
 object Commodity {
     val finder: Model.Find[Long,Commodity] = new Model.Finder[Long,Commodity](classOf[Commodity])
-    object Dao extends EbeanCrudDao[Long,Commodity](finder)
+    object Dao extends WebSocketDao[Commodity](finder)
```
Comment (aj-michael) [inline app/io/pathfinder/models/Commodity.scala:13]: Why is this dao specific to web sockets? I don't think that the data layer should be coupled to the transport layer.

Comment (danhanson) [inline app/io/pathfinder/models/Commodity.scala:13]: It is for push requests, the websocket dao is intended to push notifications to subscribed websockets. Anyone using the dao to access its data doesn't need to worry about that though

**Comment by aj-michael** [comment]
LGTM

**Pull request merged**